lib/libzpool/uzfs_rebuilding.c

```diff
@@ -0,0 +1,293 @@
+		if (ret) {
+			printf("Failed to create snapshot for %s\n",
+			    zv->zv_name);
+			strfree(dataset);
+			strfree(snap_name);
+			return (ret);
+		}
+
+		ret = uzfs_open_dataset(zv->zv_spa, dataset, snap_zv);
+	}
+
+	if (ret != 0) {
+		strfree(dataset);
+		printf("Failed to own snapshot.. err(%d)\n", ret);
+		return (ret);
+	}
+
+	strfree(dataset);
+	return (ret);
+}
+
+void
+destroy_snapshot_zv(zvol_state_t *zv, char *snap_name)
+{
+	char *dataset;
+
+	dataset = kmem_asprintf("%s@%s", zv->zv_name, snap_name);
+	(void) dsl_destroy_snapshot(dataset, B_FALSE);
+	strfree(dataset);
+}
+
+int
+uzfs_get_io_diff(zvol_state_t *zv, blk_metadata_t *low,
+    uzfs_get_io_diff_cb_t *func, off_t lun_offset, size_t lun_len, void *arg)
+{
+	uint64_t blocksize = zv->zv_volmetablocksize;
+	uint64_t metadata_read_chunk_size = 10 * blocksize;
+	uint64_t metaobjectsize = (zv->zv_volsize / zv->zv_metavolblocksize) *
+	    zv->zv_volmetadatasize;
+	uint64_t metadatasize = zv->zv_volmetadatasize;
+	char *buf, *snap_name;
+	uint64_t i, read;
+	uint64_t offset, len, end;
+	int ret = 0;
+	int diff_count = 0, last_index = 0;
+	uint64_t last_lun_offset = 0;
+	blk_metadata_t *last_md;
+	zvol_state_t *snap_zv;
+	metaobj_blk_offset_t snap_metablk;
+
+	if (!func || (lun_offset + lun_len) > zv->zv_volsize)
+		return (EINVAL);
+
+	get_zv_metaobj_block_details(&snap_metablk, zv, lun_offset, lun_len);
+	offset = snap_metablk.m_offset;
+	end = snap_metablk.m_offset + snap_metablk.m_len;
+
+	if (end > metaobjectsize)
+		end = metaobjectsize;
+
+	snap_name = kmem_asprintf("%s%llu", IO_DIFF_SNAPNAME, low->io_num);
+
+	ret = get_snapshot_zv(zv, snap_name, &snap_zv);
+	if (ret != 0) {
+		printf("failed to get snapshot info for %s io_num:%lu\n",
+		    zv->zv_name, low->io_num);
+		strfree(snap_name);
+		return (ret);
+	}
+
+	metadata_read_chunk_size = (metadata_read_chunk_size / metadatasize) *
+	    metadatasize;
+	buf = umem_alloc(metadata_read_chunk_size, KM_SLEEP);
+	len = metadata_read_chunk_size;
+
+	for (; offset < end && !ret; offset += len) {
+		read = 0;
+		len = metadata_read_chunk_size;
+
+		if ((offset + len) > end)
+			len = (end - offset);
+
+		ret = uzfs_read_metadata(snap_zv, buf, offset, len, &read);
+
+		if (read != len || ret)
+			break;
+
+		lun_offset = (offset / metadatasize) * zv->zv_metavolblocksize;
+		for (i = 0; i < len && !ret; i += sizeof (blk_metadata_t)) {
+			if (!iszero((blk_metadata_t *)(buf+i)) &&
+			    (compare_blk_metadata((blk_metadata_t *)(buf + i),
+			    low) > 0)) {
+				/*
+				 * We will keep track of last lun_offset having
+				 * metadata lesser than incoming_metadata and
+				 * join adjacent chunk with the same on_disk
+				 * io_number.
+				 */
+				if (diff_count == 0) {
+					last_lun_offset = lun_offset;
+					last_md = (blk_metadata_t *)(buf+i);
+					last_index = i;
+				}
+
+				// Increase diff_count if on_disk io number is
+				// same as last one.
+				diff_count++;
+
+				if (compare_blk_metadata((blk_metadata_t *)
```

Comment:
diff_count should go into 'else' case, and, last_lun_offset/last_md/last_index need to be set in 'if' after EXECUTE_... macro

Reply:
ok